app/models/registry.rb

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env ruby
+
+
+class Registry
+  class NoBearerRealmException < RuntimeError
+  end
+
+  class AuthorizationError < RuntimeError
+  end
+
+  class ManifestNotFoundError < RuntimeError
+  end
+
+  class CredentialsMissingError < RuntimeError
+  end
+
+  def initialize(host, use_ssl=true, username=nil, password=nil)
+    @host = host
+    @use_ssl = use_ssl
+    @base_url = "http#{'s' if @use_ssl}://#{@host}/v2/"
+    @username = username
+    @password = password
+  end
+
+  def has_credentials?
+    @username && @password
+  end
+
+  def manifest(repository, tag='latest')
+    res = get_request("#{repository}/manifests/#{tag}")
+    if res.code.to_i == 200
+      JSON.parse(res.body)
+    elsif res.code.to_i == 404
+      fail ManifestNotFoundError, "Cannot find manifest for #{repository}:#{tag}"
+    else
+      fail "Something went wrong while fetching manifest for #{repository}:#{tag}:" \
+        "[#{res.code}] - #{res.body}"
+    end
+  end
+
+  def get_request(path, request_auth_token=true)
+    uri = URI.join(@base_url, path)
+    req = Net::HTTP::Get.new(uri)
+    req['Authorization'] = "Bearer #{@token}" if @token
+
+    res = Net::HTTP.start(
+      uri.hostname,
+      uri.port,
+      use_ssl: uri.scheme == 'https') do |http|
+      http.request(req)
+    end
+
+    if res.code.to_i == 401
+      if request_auth_token
+        request_auth_token(res)
+        return get_request(path, false)
+      else
+        fail AuthorizationError, "Cannot obtain authentication token, check your username and password: #{res}"
+      end
+    else
+      res
+    end
+  end
+
+  private
+
+  def request_auth_token(unhauthorized_response)
+    bearer_realm, query = parse_unhauthorized_response(unhauthorized_response)
+
+    uri = URI("#{bearer_realm}?#{query.to_query}")
+
+    req = Net::HTTP::Get.new(uri)
+    req.basic_auth(@username, @password) if has_credentials?
+
+    res = Net::HTTP.start(
```

Comment:
Same here better to use solid solution as this is a golden stardard in Ruby more or less